### PR TITLE
all: Modernize error wrapping

### PR DESCRIPTION
**Pull request by calmh**
This replaces old style errors.Wrap with modern fmt.Errorf and removes
the (direct) dependency on github.com/pkg/errors. A couple of cases are
adjusted by hand as previously errors.Wrap(nil, ...) would return nil,
which is not what fmt.Errorf does.

Existing calls to errors.New, errors.Is and errors.As work the same as previously.

### CODE REVIEW

**Comment by acolomb** [comment]
Currently short on spare time, hopefully I can take a look at this over the weekend.

File: lib/model/folder.go [hunk 1]
```diff
@@ -317,7 +316,7 @@ func (f *folder) getHealthErrorAndLoadIgnores() error {
 	}
 	if f.Type != config.FolderTypeReceiveEncrypted {
 		if err := f.ignores.Load(".stignore"); err != nil && !fs.IsNotExist(err) {
-			return errors.Wrap(err, "loading ignores")
+			return fmt.Errorf("loading ignores: %w", err)
```
Comment (rasa) [inline lib/model/folder.go:319]: Should this and others be changed from `loading` to simply `load` to be consistent?

Comment (acolomb) [reply lib/model/folder.go:319]: While we're at it, yeah I'd prefer that.

Comment (acolomb) [inline lib/model/folder.go:319]:
```suggestion
			return fmt.Errorf("load ignores: %w", err)
```

File: lib/model/folder_sendrecv.go [hunk 2]
```diff
@@ -593,7 +593,7 @@ func (f *sendReceiveFolder) handleDir(file protocol.FileInfo, snap *db.Snapshot,
 		// Check that it is what we have in the database.
 		curFile, hasCurFile := snap.Get(protocol.LocalDeviceID, file.Name)
 		if err := f.scanIfItemChanged(file.Name, info, curFile, hasCurFile, scanChan); err != nil {
-			err = errors.Wrap(err, "handling dir")
+			err = fmt.Errorf("handling dir: %w", err)
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:596]:
```suggestion
			err = fmt.Errorf("handle dir: %w", err)
```

File: lib/model/folder_sendrecv.go [hunk 3]
```diff
@@ -647,13 +647,13 @@ func (f *sendReceiveFolder) handleDir(file protocol.FileInfo, snap *db.Snapshot,
 		if err = f.inWritableDir(mkdir, file.Name); err == nil {
 			dbUpdateChan <- dbUpdateJob{file, dbUpdateHandleDir}
 		} else {
-			f.newPullError(file.Name, errors.Wrap(err, "creating directory"))
+			f.newPullError(file.Name, fmt.Errorf("creating directory: %w", err))
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:650]:
```suggestion
			f.newPullError(file.Name, fmt.Errorf("create directory: %w", err))
```

File: lib/model/folder_sendrecv.go [hunk 4]
```diff
@@ -647,13 +647,13 @@ func (f *sendReceiveFolder) handleDir(file protocol.FileInfo, snap *db.Snapshot,
...
 		}
 		return
 	// Weird error when stat()'ing the dir. Probably won't work to do
 	// anything else with it if we can't even stat() it.
 	case err != nil:
-		f.newPullError(file.Name, errors.Wrap(err, "checking file to be replaced"))
+		f.newPullError(file.Name, fmt.Errorf("checking file to be replaced: %w", err))
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:656]:
```suggestion
		f.newPullError(file.Name, fmt.Errorf("check file to be replaced: %w", err))
```

File: lib/model/folder_sendrecv.go [hunk 5]
```diff
@@ -647,13 +647,13 @@ func (f *sendReceiveFolder) handleDir(file protocol.FileInfo, snap *db.Snapshot,
 		if err = f.inWritableDir(mkdir, file.Name); err == nil {
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:647]: Invert to `err != nil` for consistency?

File: lib/model/folder_sendrecv.go [hunk 6]
```diff
@@ -675,7 +675,7 @@ func (f *sendReceiveFolder) checkParent(file string, scanChan chan<- string) boo
 	parent := filepath.Dir(file)
 
 	if err := osutil.TraversesSymlink(f.mtimefs, parent); err != nil {
-		f.newPullError(file, errors.Wrap(err, "checking parent dirs"))
+		f.newPullError(file, fmt.Errorf("checking parent dirs: %w", err))
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:678]:
```suggestion
		f.newPullError(file, fmt.Errorf("check parent dirs: %w", err))
```

File: lib/model/folder_sendrecv.go [hunk 7]
```diff
@@ -700,7 +700,7 @@ func (f *sendReceiveFolder) checkParent(file string, scanChan chan<- string) boo
 	}
 	l.Debugf("%v creating parent directory of %v", f, file)
 	if err := f.mtimefs.MkdirAll(parent, 0755); err != nil {
-		f.newPullError(file, errors.Wrap(err, "creating parent dir"))
+		f.newPullError(file, fmt.Errorf("creating parent dir: %w", err))
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:703]:
```suggestion
		f.newPullError(file, fmt.Errorf("create parent dir: %w", err))
```

File: lib/model/folder_sendrecv.go [hunk 8]
```diff
@@ -761,7 +761,7 @@ func (f *sendReceiveFolder) handleSymlink(file protocol.FileInfo, snap *db.Snaps
 	if err = f.inWritableDir(createLink, file.Name); err == nil {
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:761]: Invert to `err != nil` for consistency?

File: lib/model/folder_sendrecv.go [hunk 9]
```diff
@@ -1307,7 +1307,7 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 		for _, block := range state.blocks {
 			select {
 			case <-f.ctx.Done():
-				state.fail(errors.Wrap(f.ctx.Err(), "folder stopped"))
+				state.fail(fmt.Errorf("folder stopped: %w", f.ctx.Err()))
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:1310]: Are we sure that `Err()` will be non-nil?  Comment might help in that case.

File: lib/model/folder_sendrecv.go [hunk 10]
```diff
@@ -1536,7 +1536,7 @@ loop:
 	for {
 		select {
 		case <-f.ctx.Done():
-			state.fail(errors.Wrap(f.ctx.Err(), "folder stopped"))
+			state.fail(fmt.Errorf("folder stopped: %w", f.ctx.Err()))
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:1539]: Are we sure that `Err()` will be non-nil?  Comment might help in that case.

File: lib/model/folder_sendrecv.go [hunk 11]
```diff
@@ -1888,7 +1888,9 @@ func (f *sendReceiveFolder) newPullError(path string, err error) {
 // deleteItemOnDisk deletes the file represented by old that is about to be replaced by new.
 func (f *sendReceiveFolder) deleteItemOnDisk(item protocol.FileInfo, snap *db.Snapshot, scanChan chan<- string) (err error) {
 	defer func() {
-		err = errors.Wrap(err, contextRemovingOldItem)
+		if err != nil {
+			err = errors.Wrap(err, contextRemovingOldItem)
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:1892]:
```suggestion
			err = return fmt.Errorf("%s: %w", contextRemovingOldItem, err)
```

File: lib/model/folder_sendrecv.go [hunk 12]
```diff
@@ -2056,7 +2058,7 @@ func (f *sendReceiveFolder) scanIfItemChanged(name string, stat fs.FileInfo, ite
 	// touching the item.
 	statItem, err := scanner.CreateFileInfo(stat, item.Name, f.mtimefs)
 	if err != nil {
-		return errors.Wrap(err, "comparing item on disk to db")
+		return fmt.Errorf("comparing item on disk to db: %w", err)
```
Comment (acolomb) [inline lib/model/folder_sendrecv.go:2061]:
```suggestion
		return fmt.Errorf("compare item on disk to db: %w", err)
```

File: lib/model/sharedpullerstate.go [hunk 13]
```diff
@@ -177,12 +176,12 @@ func (s *sharedPullerState) tempFileInWritableDir(_ string) error {
 		// what the umask dictates.
 
 		if err := s.fs.Chmod(s.tempName, mode); err != nil {
-			return errors.Wrap(err, "setting perms on temp file")
+			return fmt.Errorf("setting perms on temp file: %w", err)
```
Comment (acolomb) [inline lib/model/sharedpullerstate.go:179]:
```suggestion
			return fmt.Errorf("set perms on temp file: %w", err)
```

File: lib/model/sharedpullerstate.go [hunk 14]
```diff
@@ -177,12 +176,12 @@ func (s *sharedPullerState) tempFileInWritableDir(_ string) error {
...
 		}
 	}
 	fd, err := s.fs.OpenFile(s.tempName, flags, mode)
 	if err != nil {
-		return errors.Wrap(err, "opening temp file")
+		return fmt.Errorf("opening temp file: %w", err)
```
Comment (acolomb) [inline lib/model/sharedpullerstate.go:184]:
```suggestion
		return fmt.Errorf("open temp file: %w", err)
```

File: lib/syncthing/utils.go [hunk 15]
```diff
@@ -93,17 +92,17 @@ func LoadConfigAtStartup(path string, cert tls.Certificate, evLogger events.Logg
 	if fs.IsNotExist(err) {
 		cfg, err = DefaultConfig(path, myID, evLogger, noDefaultFolder, skipPortProbing)
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to generate default config")
+			return nil, fmt.Errorf("failed to generate default config: %w", err)
```
Comment (acolomb) [inline lib/syncthing/utils.go:95]:
```suggestion
			return nil, fmt.Errorf("generate default config: %w", err)
```

File: lib/syncthing/utils.go [hunk 16]
```diff
@@ -93,17 +92,17 @@ func LoadConfigAtStartup(path string, cert tls.Certificate, evLogger events.Logg
...
 		}
 		err = cfg.Save()
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to save default config")
+			return nil, fmt.Errorf("failed to save default config: %w", err)
```
Comment (acolomb) [inline lib/syncthing/utils.go:99]:
```suggestion
			return nil, fmt.Errorf("save default config: %w", err)
```

File: lib/syncthing/utils.go [hunk 17]
```diff
@@ -93,17 +92,17 @@ func LoadConfigAtStartup(path string, cert tls.Certificate, evLogger events.Logg
...
 		}
 		l.Infof("Default config saved. Edit %s to taste (with Syncthing stopped) or use the GUI", cfg.ConfigPath())
 	} else if err == io.EOF {
 		return nil, errors.New("failed to load config: unexpected end of file. Truncated or empty configuration?")
 	} else if err != nil {
-		return nil, errors.Wrap(err, "failed to load config")
+		return nil, fmt.Errorf("failed to load config: %w", err)
```
Comment (acolomb) [inline lib/syncthing/utils.go:105]:
```suggestion
		return nil, fmt.Errorf("load config: %w", err)
```

File: lib/syncthing/utils.go [hunk 18]
```diff
@@ -93,17 +92,17 @@ func LoadConfigAtStartup(path string, cert tls.Certificate, evLogger events.Logg
...
+			return nil, fmt.Errorf("failed to generate default config: %w", err)
 		}
 		err = cfg.Save()
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to save default config")
+			return nil, fmt.Errorf("failed to save default config: %w", err)
 		}
 		l.Infof("Default config saved. Edit %s to taste (with Syncthing stopped) or use the GUI", cfg.ConfigPath())
 	} else if err == io.EOF {
 		return nil, errors.New("failed to load config: unexpected end of file. Truncated or empty configuration?")
```
Comment (acolomb) [inline lib/syncthing/utils.go:103]:
```suggestion
		return nil, errors.New("load config: unexpected end of file. Truncated or empty configuration?")
```

**Review by acolomb — CHANGES_REQUESTED** [review]
Looks good in general.  Some tidying up we could do along the way suggested.

File: lib/model/folder.go (hunk 1 above)
Comment (calmh) [reply lib/model/folder.go:319]: Yeah, no, this is about the way errors are generated. I'd welcome a review of returned errors (and have a number of changes queued that add additional context strings), but this PR is not that.

File: lib/model/folder_sendrecv.go (hunk 10 above)
Comment (calmh) [reply lib/model/folder_sendrecv.go:1539]: We are; context docs say "// If Done is closed, Err returns a non-nil error explaining why:". I don't think really requires explanation, it wouldn't be a question on anyone's mind unless the old Wrap() function had the odd-ish feature of wrapping nil as nil.

**Comment by calmh** [comment]
Thanks for the review and catching the one place I apparently missed an Wrap()... But I really *don't* want to "clean up" all the error strings or invert logic in places etc; this is about making a straight change to a modern API. If we want to refactor all error strings that's fine, but it's a separate project and certainly should affect places that *weren't* errors.Wrap previously..

**Review by acolomb — APPROVED** [review]

**Pull request merged**